### ISSUE TITLE
IAuthenticationProvider is deprecated? 

### ISSUE THREAD

**Issue by iamkarlson** [microsoftgraph/msgraph-sdk-java#545]
Does anybody actually using this? if you just follow the steps from the readme half of it is either deprecated or archived. What is a "normal' way of using MS graph from Java? Does anyone from MS can update docs as how you suppose we have to deal with it?

**Comment by baywet** [comment]
Hi @iamkarlson 
Thanks for reaching out.
We're working on a major revamp for [version 3](https://github.com/microsoftgraph/msgraph-sdk-java/milestone/4)
There's currently two issues with the library:
1. A number of classes and interfaces should be in the core library instead of this generated library
1. There's a duplication in the authentication stack (see ICoreAuthenticationProvider and IAuthenticationProvider)

We're working to streamline the experience and provide one clean API surface, once this is done, we'll make sure we update documentation and samples.
For now, no action is required on your end besides making sure you implement ICoreAuthenticationProvider as well.

Let me know if you have further questions.

**Comment by iamkarlson** [comment]
@baywet Thanks for explanation on what's going on in your project and what would be in the _future_. However, I need to use it right now and I don't want to bother with you implementation detail. My point is that as Microsoft it seems very unprofessional to have such a mess with SDK for one of the flagship products (azure). You could make your redesign and revamp in a separate branch with all the process in place. This would allow you to concentrate on the work you're doing and still keep a working version of the SDK. As for now I decided not to use SDK at all because the overall quality is unacceptable for production use and that would mean that I probably will never use SDK for that project. I cannot be sure but quite confident that I'm not alone in such thoughts. 

Thanks! Good luck with you rework!

**Comment by baywet** [comment]
@iamkarlson thanks for the feedback. The goal of that annotation was to give people a heads up on the upcoming work, but I can see it can be confusing. And this is not implementation detail as you might choose to use our authentication library or implement your own code at this step.

Besides this confusing annotation, I'd love to hear what additional feedback you have for the SDK so we can improve it. We've already fixed 90+ bugs in the SDK over the past few months to drastically improve the overall quality.
If there are other issues that haven't been captured yet, please open those with detailed information, we'll be happy to address those and improve the overall experience.

**Comment by hosmanagic-axcient** [comment]
> For now, no action is required on your end besides making sure you implement ICoreAuthenticationProvider **as well**.
I believe those are the keywords. It's not possible to simply replace `IAuthenticationProvider` with `ICoreAuthenticationProvider`, as `IGraphServiceClient` still depends on the deprecated interface.

> Besides this confusing annotation, I'd love to hear what additional feedback you have for the SDK so we can improve it.
I know this was addressed to @iamkarlson , but here's my 2 cents:
1. The unusual inheritance of request and request builder classes. I'll try to provide some concrete examples.
2. The raw JSON object attached to each "model" class (Message, Event etc.). It makes the model more complex and coupled with the implementation. Ideally, IMHO, we have a simple model class. All (de-)serialization details are handled by the Graph client. This also decreases the memory footprint by a lot.
3. It'd be nice if the SDK worked on a higher level, or, at least, provided a high-level API, e.g.:
```java
graph.users("user-id").messages("message-id").delete(); // or send(), or moveTo()
graph.users("user-id").messages("message-id").get("list of fields or relationships goes here"); // or send(), or moveTo()
```
That'd provide for more concise syntax, and I wouldn't need to think about the low-level stuff.

**Comment by baywet** [comment]
Thanks a lot for that feedback, it help shape the future of the SDK!

1. Ok, I'll wait for the examples. :)
2. I'm not sure what you mean by "raw JSON attached to model". Are you suggesting that you need access to the underlying JSON from the response?
3. I believe the "current way" of expressing those requests with the SDK today are:
     - `graph.users("user-id").messages("message-id").buildRequest().delete();` are you saying the build request part makes it "too long"? 
     - `graph.users("user-id").messages("message-id").buildRequest().select("list of fields or relationships goes here").get();`. We are following the odata semantics to make it easy for someone to translate from the API docs/Graph explorer to Java code. Is the select here "too long" as well? In that case in your example, what would you do for orderby, expand, top, filter... ?
4. This is what's being worked on here https://github.com/microsoftgraph/msgraph-sdk-java-core/pull/42. However some consumers still want to use their own auth libraries or provide the token themselves from their execution context.

**Comment by hosmanagic-axcient** [comment]
2. Actually, the opposite is what I'm talking about. From what I know and can see atm, all SDK "entities" have a raw JSON objects attached to them, for example [this](https://github.com/microsoftgraph/msgraph-sdk-java/blob/d29fdc3a3be3d428d4c0cf78d8a1815dc1b53529/src/main/java/com/microsoft/graph/models/extensions/OutlookItem.java#L62). That doubles the memory consumption for "models" (not sure what's the correct SDK terminology for that).

3. > We are following the odata semantics

That's what I think makes it too verbose. I'm basically suggesting making the SDK's syntax more "domain driven". I'd like to write my code using a syntax/language which is completely Microsoft (Azure, Exchange, OneDrive, SharePoint) terminology, and not HTTP/OData.

In other words, right now, using the SDK feels like constructing a low-level HTTP call using `curl`. Maybe another example:
```java
graph.users(userId)
                .mailFolders(mailFolderId)
                .messages()
                .buildRequest()
                .post(msg);
```
can be done in this way too:
```java
graph.users(userId)
                .mailFolders(mailFolderId)
                .save(msg);
```
4. That PR looks great! Having the ability to use a different auth. lib. or provide the token myself is great, but IMHO that's as far as it should go. Instead of me, as a SDK user, adding a token to a request, I'd like to just give the token to the SDK, e.g. instead of 
```java
@Override
    public void authenticateRequest(IHttpRequest request) {
        String authBearer = makeAuthBearer();
        request.addHeader("Authorization", authBearer);
    }
```
I'd like to be able to do this:
```java
@Override
    public String generateAuthToken() {
        return makeAuthBearer();
    }
```
This has 2 benefits:
a. The SDK is free to change the underlying HTTP layer without creating any breaking changes for the SDK users. Sometimes in future, you might to switch to, for example, Apache's HTTP Client, and you'll be able to do that very easily.
b. The above is a benefit for me, as an SDK user too. I'll be able to upgrade the SDK with such changes without changing my code.

**Comment by baywet** [comment]
2. you make a really good point, and after having a quick look at it, it seems that after all the changes I implemented for v3 the rawObject is only used for collections at a few places. Which we could easily workaround by having a field and assigning during deserialization instead. Can you open a new issue in [core](https://github.com/microsoftgraph/msgraph-sdk-java-core) specifically about that so I can loop in my PMs please?
3. Taking your example, how would you then differentiate between a save action on a message folder and on an email? the generated "code path" would be the same and conflicting.
4. I just pushed changes among these lines, please feel free to go and review the PR over there, leaving comments.

**Comment by hosmanagic-axcient** [comment]
2. Done (ftr, the issue is https://github.com/microsoftgraph/msgraph-sdk-java-core/issues/118)
3. We'd use method overrides, e.g.
```java
Message msg = createMessage()
graph.users(userId)
                .mailFolders(mailFolderId)
                .save(msg);
```    
and
```java
MailFolder mf = createMailFolder()
graph.users(userId)
                .mailFolders(mailFolderId)
                .save(mf);
```
4. Thanks, I'll be glad to do so.

**Comment by baywet** [comment]
2. Thanks!
3. The API is so large that we'd end up having conflicts. That's already a challenge we're running into today in some cases and this is why we have to be really careful to follow the API path in the code path. I can see the argument to remove the buildRequest section thought. There's ongoing experimental work along those lines. But that's not going to land any time soon.
4. Thanks!
